### PR TITLE
[MM-60689] Add Webhook ID in the logs

### PR DESCRIPTION
**Pull request by vish9812**
#### Summary

Adds webhook IDs to the logs when handling webhooks to identify which webhook caused the issue.

#### Ticket Link

JIRA - https://mattermost.atlassian.net/browse/MM-60689

#### Release Note

```release-note
NONE
```

### CODE REVIEW

**Comment by larkox** [comment]
This only covers outgoing webhooks. Can you add something like this also for incoming webhooks? I think those should be handled on `server/channels/web/webhook.go`.

File: server/channels/app/webhook.go [hunk 1]
```diff
@@ -103,7 +103,7 @@ func (a *App) TriggerWebhook(c request.CTX, payload *model.OutgoingWebhookPayloa
 		contentType = "application/json"
 		jsonBytes, err = json.Marshal(payload)
 		if err != nil {
-			c.Logger().Warn("Failed to encode to JSON", mlog.Err(err))
+			c.Logger().Warn("Failed to encode to JSON", mlog.String("outgoing_webhook_id", hook.Id), mlog.Err(err))
```
Comment (larkox) [inline server/channels/app/webhook.go:106]: I wonder if it makes sense to use for this function something like `c.Logger().With(...)`. That way, we only add the id in the with, and then we can use the newly generated logger everywhere.

0/5

File: server/i18n/en.json [hunk 2]
```diff
@@ -4896,23 +4896,23 @@
   },
   {
     "id": "app.command_webhook.get.internal_error",
-    "translation": "Unable to get the webhook."
+    "translation": "Unable to get the webhook {{.hook_id}}"
```
Comment (hanzei) [inline server/i18n/en.json:4899]: I wonder what the value of mirroring the input as part of the error is. Doesn't the caller already know the HookId? The same question applies to some of the other error messages.

File: server/channels/app/webhook.go [hunk 3]
```diff
@@ -883,17 +884,18 @@ func (a *App) CreateCommandWebhook(commandID string, args *model.CommandArgs) (*
...
 func (a *App) HandleCommandWebhook(c request.CTX, hookID string, response *model.CommandResponse) *model.AppError {
 	if response == nil {
-		return model.NewAppError("HandleCommandWebhook", "app.command_webhook.handle_command_webhook.parse", nil, "", http.StatusBadRequest)
+		return model.NewAppError("HandleCommandWebhook", "app.command_webhook.handle_command_webhook.parse", map[string]any{"hook_id": hookID}, "", http.StatusBadRequest)
 	}
 
 	hook, nErr := a.Srv().Store().CommandWebhook().Get(hookID)
 	if nErr != nil {
 		var nfErr *store.ErrNotFound
+
```
Comment (hanzei) [inline server/channels/app/webhook.go:893]: Unrelated change?

**Review by hanzei — CHANGES_REQUESTED** [review]
Nice work. I really like the logging :tada:

File: server/channels/app/webhook.go [hunk 4]
```diff
@@ -734,12 +735,12 @@ func (a *App) HandleIncomingWebhook(c request.CTX, hookID string, req *model.Inc
 	}()
 
 	if req == nil {
-		return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.parse.app_error", nil, "", http.StatusBadRequest)
+		return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.parse.app_error", map[string]any{"hook_id": hookID}, "", http.StatusBadRequest)
```
Comment (larkox) [inline server/channels/app/webhook.go:738]: This only apply if we keep the hook id in the error.

The same way as with `Logger().With()`, I wonder if it is reasonable to define a variable for this (e.g. `errorDefaultProps`, or any other better naming), so we don't have to write every time `map[string]any{"hook_id": hookID}`.

What do you think?

**Review by larkox — APPROVED** [review]
No extra comments on my side, just a small improvement if we keep the hook id in the errors.

**Comment by agnivade** [comment]
This is really great work. We absolutely need more context in our error logs. cc @lieut-data @isacikgoz from the convo re: logging earlier.

File: server/i18n/en.json (hunk 2 above)
Comment (vish9812) [reply server/i18n/en.json]: You are absolutely right. If caller has the data, then it should log the data.

File: server/channels/web/webhook.go [hunk 5]
```diff
@@ -63,12 +59,13 @@ func incomingWebhook(c *Context, w http.ResponseWriter, r *http.Request) {
...
+	errCtx["media_type"] = mediaType
 	if mediaType == "application/x-www-form-urlencoded" {
 		payload := strings.NewReader(r.FormValue("payload"))
 
-		incomingWebhookPayload, err = decodePayload(payload)
-		if err != nil {
-			c.Err = err
+		incomingWebhookPayload, appErr = decodePayload(payload)
+		if appErr != nil {
+			c.Err = model.NewAppError("incomingWebhook", "web.incoming_webhook.decode.app_error", errCtx, "", http.StatusBadRequest).Wrap(appErr)
```
Comment (hanzei) [inline server/channels/web/webhook.go:68]: Are we returning the wrapped error via the API? If not, is that something we want to keep?

Comment (hanzei) [reply server/channels/web/webhook.go:68]: Should the error message make it clear that the logs need to be checked?

**Review by hanzei — CHANGES_REQUESTED** [review]
This PR is getting better and better :tada: 

One question regarding returning errors via the API vs. logging them,

File: server/channels/app/webhook.go [hunk 6]
```diff
@@ -780,7 +781,7 @@ func (a *App) HandleIncomingWebhook(c request.CTX, hookID string, req *model.Inc
 		if channelName[0] == '@' {
 			result, nErr := a.Srv().Store().User().GetByUsername(channelName[1:])
 			if nErr != nil {
-				return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.user.app_error", nil, "", http.StatusBadRequest).Wrap(nErr)
+				return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.user.app_error", map[string]any{"user": channelName[1:]}, "", http.StatusBadRequest).Wrap(nErr)
```
Comment (lieut-data) [inline server/channels/app/webhook.go:784]: To confirm, do we still have enough context in the resulting log to trace?

File: server/channels/app/webhook.go [hunk 7]
```diff
@@ -831,16 +833,16 @@ func (a *App) HandleIncomingWebhook(c request.CTX, hookID string, req *model.Inc
...
 
 	if hook.ChannelLocked && hook.ChannelId != channel.Id {
-		return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.channel_locked.app_error", nil, "", http.StatusForbidden)
+		return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.channel_locked.app_error", map[string]any{"channel_id": channel.Id}, "", http.StatusForbidden)
 	}
 
 	resultU := <-uchan
 	if resultU.NErr != nil {
-		return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.user.app_error", nil, "", http.StatusForbidden).Wrap(resultU.NErr)
+		return model.NewAppError("HandleIncomingWebhook", "web.incoming_webhook.user.app_error", map[string]any{"user": hook.UserId}, "", http.StatusForbidden).Wrap(resultU.NErr)
```
Comment (lieut-data) [inline server/channels/app/webhook.go:841]: Should we use `user_id` here instead of `user` as the key?

**Review by lieut-data — APPROVED** [review]
A few thoughts, not strictly blocking. Thanks @vish9812!

File: server/channels/app/webhook.go (hunk 7 above)
Comment (vish9812) [reply server/channels/app/webhook.go:841]: This one is a special case, where the same error "can't find the user" is used at 2 search attempts, one by user_id and another by username.
So, went with the `user` key.

File: server/channels/app/webhook.go (hunk 6 above)
Comment (vish9812) [reply server/channels/app/webhook.go:784]: Yes, this specific log provides the context used when calling the above function. And the _above_ layer(API) will provide further context.

File: server/channels/web/webhook.go (hunk 5 above)
Comment (vish9812) [reply server/channels/web/webhook.go:68]: - Yes, wrapped error does show in the API response. So, we should keep it.
- Error messages telling to check error logs, seems a bit too much to me.

File: server/channels/app/webhook.go (hunk 7 above)
Comment (lieut-data) [reply server/channels/app/webhook.go:842]: I don't think the property name should match the error message, but describe the field being logged. Is this always guaranteed to be a `user_id`?

Comment (lieut-data) [reply server/channels/app/webhook.go:842]: (Consider the typical use case of scrubbing logs with Loki, and filtering on `user_id=XXX`. It's ideal to have this find all related results, and not have to remember to do `user=XXX` for cases like this.)

Comment (vish9812) [reply server/channels/app/webhook.go:842]: Actually, my intention wasn’t to match the error message exactly but rather to account for both cases within a single error message. 
In this implementation, it’s not always guaranteed to be `user_id`; in one case, we’re passing `username`, and in another, `user_id` to the _same_ error message.

For the log filtering, it would be even better to standardize with a generic key like `user`. This way, whether the search term is `user_id` or `username`, we could simply filter by consistently using `user=xxx`.
Also, I'm not sure how the Loki search works, but couldn't you simply search for `xxx` or `=xxx` no matter the key?

Comment (lieut-data) [reply server/channels/app/webhook.go:842]: > For the log filtering, it would be even better to standardize with a generic key like user.

I think the momentum is already behind naming the keys explicitly. But in this case, if it's sometimes both, then it's probably fine :)

Comment (agnivade) [reply server/channels/app/webhook.go:842]: > Also, I'm not sure how the Loki search works, but couldn't you simply search for xxx or =xxx no matter the key?

That is not idiomatic way to use Loki, although yes, it's doable. The idiomatic way is to use structured logging, and filter with the keys. This is much more powerful, and you can use Loki in-built utilities to filter for the value of a given `user_id` key, do aggregation operations, and so on.

This is unrelated to this specific case where the same error is being used for username and user_id both. But just a general idea on idiomatic logging for better usage of Loki.

**Review by hanzei — APPROVED** [review]
LGTM :+1:
